Add Oscilloscope Effect

regulus79 opened the pull request:
This PR adds an effect to visually show the incoming waveform. This is similar to the oscilloscope at the top tool bar, but it has many more features, including pausing the display, scaling the amplitude, zooming in certain sections, and viewing the stereo waveform.

The audio processing portion simply saves the incoming buffer into a ring buffer. Most of the code is for the gui, which includes a zoomable+scrollable waveform graph to display the contents of the ring buffer.

This PR does NOT include a pitch tracking option to lock on to one period of the waveform. Perhaps that can be implemented in the future.

# Demo

https://github.com/user-attachments/assets/34f36e90-713d-455b-91de-8f728df0c045


![image](https://github.com/user-attachments/assets/69ea6102-dc11-46fa-a56c-a2c2c6f778e4)

![image](https://github.com/user-attachments/assets/cfc6151e-da7b-4df5-ba9a-4059bf3f6337)

regulus79 committed:
Initial commit

regulus79 committed:
Update everything, remove pitch tracking

regulus79 committed:
Add stereo and fix zoom bug

regulus79 committed:
Remove Effect from name, and fix hardcoded color issues

regulus79 committed:
Update copyright

regulus79 committed:
Attempt to fix spacing issues

regulus79 committed:
Clean up

regulus79 committed:
Attempt to fix spacing again

regulus79 committed:
Move window zero point to right, fix graphical bug, and rework stuff to be cleaner

regulus79 committed:
Update classic style.css

regulus79 committed:
Attempt to fix CodeFactor warnings

regulus79 committed:
Flip oscilloscope and slightly fix/optimize drawing

regulus79 committed:
Initial min/max sample drawing

regulus79 committed:
Fix flickering

regulus79 committed:
Fix orientation

regulus79 committed:
Fix start index quantization

regulus79 committed:
Change buffer size to 3 seconds

regulus79 committed:
Make window size logarithmic

regulus79 commented:
Should be ready to test!

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,216 @@
+				p.drawLine(
+					width() * f / (windowSize - framesPerPixel),
+					height() * (1 - buffer[currentIndex].left() * amp) / 2,
+					width() * (f + framesPerPixel) / (windowSize - framesPerPixel),
+					height() * (1 - buffer[nextIndex].left() * amp) / 2
+				);
+			}
+		}
+		p.setPen(m_rightColor);
+		for (int f = 0; f < windowSize - framesPerPixel; f += framesPerPixel)
+		{
+			int currentIndex = (windowStartIndex + f) % bufferSize;
+			int nextIndex = (windowStartIndex + f + framesPerPixel) % bufferSize;
+			if (framesPerPixel > 1)
+			{
+				float maxValue = buffer[currentIndex].right();
+				float minValue = maxValue;
+				for (int i = currentIndex + 1; i <= nextIndex; ++i)
+				{
+					maxValue = std::max(maxValue, buffer[i].right());
+					minValue = std::min(minValue, buffer[i].right());
+				}
+				p.drawLine(
+					width() * f / (windowSize - framesPerPixel),
+					height() * (1 - minValue * amp) / 2,
+					width() * f / (windowSize - framesPerPixel),
+					height() * (1 - maxValue * amp) / 2
+				);
+			}
+			else
+			{
+				p.drawLine(
+					width() * f / (windowSize - framesPerPixel),
+					height() * (1 - buffer[currentIndex].right() * amp) / 2,
+					width() * (f + framesPerPixel) / (windowSize - framesPerPixel),
+					height() * (1 - buffer[nextIndex].right() * amp) / 2
+				);
+			}
+		}
+	}
```
```suggestion
	const bool hq = framesPerPixel > 1;
	const bool xoffset = !hq * framesPerPixel;
	const float xscale = 1.f / (windowSize - framesPerPixel);
	const bool stereo = m_controls->m_stereoModel.value();
	if (!stereo)
	{
		p.setPen(m_monoColor);
		for (int f = 0; f < windowSize - framesPerPixel; f += framesPerPixel)
		{
			const int currentIndex = (windowStartIndex + f) % bufferSize;
			const int nextIndex = (windowStartIndex + f + framesPerPixel) % bufferSize;
			float maxValue = buffer[hq ? currentIndex : nextIndex].average();
			float minValue = buffer[currentIndex].average();
			for (int i = currentIndex + 1; hq && i <= nextIndex; ++i)
			{
				maxValue = std::max(maxValue, buffer[i].average());
				minValue = std::min(minValue, buffer[i].average());
			}
			p.drawLine(
				width() * f * xscale,
				height() * (1 - minValue * amp) / 2,
				width() * (f + xoffset) * xscale,
				height() * (1 - maxValue * amp) / 2
			);
		}
	}
	else
	{
		p.setPen(m_leftColor);
		for (int f = 0; f < windowSize - framesPerPixel; f += framesPerPixel)
		{
			const int currentIndex = (windowStartIndex + f) % bufferSize;
			const int nextIndex = (windowStartIndex + f + framesPerPixel) % bufferSize;
			float maxValue = buffer[hq ? currentIndex : nextIndex].left();
			float minValue = buffer[currentIndex].left();
			for (int i = currentIndex + 1; hq && i <= nextIndex; ++i)
			{
				maxValue = std::max(maxValue, buffer[i].left());
				minValue = std::min(minValue, buffer[i].left());
			}
			p.drawLine(
				width() * f * xscale,
				height() * (1 - minValue * amp) / 2,
				width() * (f + xoffset) * xscale,
				height() * (1 - maxValue * amp) / 2
			);
		}
		p.setPen(m_rightColor);
		for (int f = 0; f < windowSize - framesPerPixel; f += framesPerPixel)
		{
			const int currentIndex = (windowStartIndex + f) % bufferSize;
			const int nextIndex = (windowStartIndex + f + framesPerPixel) % bufferSize;
			float maxValue = buffer[hq ? currentIndex : nextIndex].right();
			float minValue = buffer[currentIndex].right();
			for (int i = currentIndex + 1; hq && i <= nextIndex; ++i)
			{
				maxValue = std::max(maxValue, buffer[i].right());
				minValue = std::min(minValue, buffer[i].right());
			}
			p.drawLine(
				width() * f * xscale,
				height() * (1 - minValue * amp) / 2,
				width() * (f + xoffset) * xscale,
				height() * (1 - maxValue * amp) / 2
			);
		}
	}
```

The code paths for when `framesPerPixel > 1` or not can be combined by using `minValue` and `maxValue` regardless, and calculating their values depending on that condition instead.

I'd suspect there's a way to also combine the mono, left, and right logic into a lambda that gets called either once or twice, since the only differences are the colors and whether or not it calls `.average()`, `.left()`, or `.right()` on the `SampleFrame`, but I don't have an example for that right now lol

regulus79 committed:
Update plugins/Oscilloscope/OscilloscopeGraph.cpp

Co-authored-by: Fawn <rubiefawn@gmail.com>

bratpeki commented:
This is great! I'd love to learn more about "Scale" and will approve after that.

regulus79 committed:
Merge branch 'master' into oscilloscope

regulus79 commented:
> if you copied the code from the Oscilloscope we already have, maybe we can reuse some of it.

I didn't copy anything from the current Oscilloscope in this PR.

regulus79 committed:
Merge branch 'master' into oscilloscope

regulus79 committed:
Use lambdas for drawing each channel

regulus79 committed:
Prevent user from accidentally zooming to reveal write position

regulus79 committed:
Initialize variables to fix warning

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,164 @@
+	{
+		p.setPen(color);
+		float minValue = 0.0f, maxValue = 0.0f;
+		for (int f = 0; f < windowSize - framesPerPixel; f += framesPerPixel)
+		{
+			const int currentIndex = (windowStartIndex + f) % bufferSize;
+			const int nextIndex = (windowStartIndex + f + framesPerPixel) % bufferSize;
+
+			if (left) { maxValue = buffer[hq ? currentIndex : nextIndex].left(); }
+			else if (right) { maxValue = buffer[hq ? currentIndex : nextIndex].right(); }
+			else if (average) { maxValue = buffer[hq ? currentIndex : nextIndex].average(); }
+
+			if (left) { minValue = buffer[currentIndex].left(); }
+			else if (right) { minValue = buffer[currentIndex].right(); }
+			else if (average) { minValue = buffer[currentIndex].average(); }
+
+			for (int i = currentIndex + 1; hq && i <= nextIndex; ++i)
+			{
+				maxValue = std::max(maxValue, buffer[i].average());
+				minValue = std::min(minValue, buffer[i].average());
+			}
+
+			p.drawLine(
+				width() * f * xscale,
+				height() * (1 - minValue * amp) / 2,
+				width() * (f + xoffset) * xscale,
+				height() * (1 - maxValue * amp) / 2
+			);
+		}
+	};
+
+	if (!stereo)
+	{
+		drawWaveform(m_monoColor, false, false, true);
+	}
+	else
+	{
+		drawWaveform(m_leftColor, true, false, false);
+		drawWaveform(m_rightColor, false, true, false);
+	}
```
I'm away from a computer that can compile to test this out, but here is an alternative that removes all the boolean parameters. I'm not sure how smart the compiler will be about the lambdas, so it's possible this suggestion is a strict downgrade depending on function call overhead. Putting it here anyway lol. I'll update this comment with more information when I get the chance to test this suggestion

```c++
// #include <functional> or replace std::function<> with C-style function pointer
auto drawWaveform = [&](const QColor& color, std::function<float(const SampleFrame&)> hmmm)
{
	p.setPen(color);
	for (int f = 0; f < windowSize - framesPerPixel; f += framesPerPixel)
	{
		const int currentIndex = (windowStartIndex + f) % bufferSize;
		const int nextIndex = (windowStartIndex + f + framesPerPixel) % bufferSize;

		float minValue = hmmm(buffer[hq ? currentIndex : nextIndex]);
		float maxValue = hmmm(buffer[currentIndex]);

		for (int i = currentIndex + 1; hq && i <= nextIndex; ++i)
		{
			maxValue = std::max(maxValue, hmmm(buffer[i]));
			minValue = std::min(minValue, hmmm(buffer[i]));
		}

		p.drawLine(
			width() * f * xscale,
			height() * (1 - minValue * amp) / 2,
			width() * (f + xoffset) * xscale,
			height() * (1 - maxValue * amp) / 2
		);
	}
};

if (!stereo)
{
	drawWaveform(m_monoColor, [](const SampleFrame& s) { return s.average(); });
}
else
{
	drawWaveform(m_leftColor, [](const SampleFrame& s) { return s.left(); });
	drawWaveform(m_rightColor, [](const SampleFrame& s) { return s.right(); });
}
```

regulus79 replied:
Ooo, that's much nicer! I've never really played with passing functions as arguments before so this is cool!

regulus79 replied:
Yayy, that brought the code complexity down just enough for CodeFactor to pass :tada: thanks!

regulus79 committed:
Pass channel/average selector function as argument to waveform drawing lambda

RainbowShatter commented:
Hey, uh, I have a Recommendation: Maybe we can add a Theme Editor next to the "Stereo" toggle button, to change the white line to other colors, what Y'all think about it?

Probably would be easy to turn the RGBA values into variables, so that each color wouldn't be locked to a specific preset.

I barely know about Qt CSS, but I'm always here to help wih anything! :D

regulus79 commented:
I made an effort to make sure all the colors are themeable, so normal themes which customize the `style.css` should be able to change the colors just fine. They are not hardcoded to a specific preset.

Changing the color of the waveform right there in the plugin is interesting--It's probably not *too* hard to implement, but that would also bring up the question of whether all the other plugins such as the equalizer and spectrum analyzer should be customizable right in the app. It would be cool to have an in-app theme editor someday, but that may be a difficult task.

Also now that we are bringing up the stereo color, I do apologize--the way I'm drawing the waveforms means I do the left channel first, then the right (or the other way I forgot), which means when the two channels overlap, the channel drawn last looks brighter. I'm still thinking if there's an easy way to fix it without like changing the pen color multiple times per line. I'm not sure, but it's probably not too big a deal.

rubiefawn commented:
> the way I'm drawing the waveforms means I do the left channel first, then the right (or the other way I forgot), which means when the two channels overlap, the channel drawn last looks brighter.

I'm sure Qt has an additive blend mode (`QPainter::CompositionMode_Plus`?)

regulus79 committed:
Fix drawing of stereo colors

regulus79 commented:
> I'm sure Qt has an additive blend mode (QPainter::CompositionMode_Plus?)

Oh thanks, I tried that and now it seems to look much better!

bratpeki reviewed (approved):
The one-off error is inapparent, IMO, so this is all good from a tester perspective! Approving, a follow-up PR in the future would be good if we find the cause.

bratpeki commented on plugins/Oscilloscope/Oscilloscope.cpp:
```diff
@@ -0,0 +1,88 @@
+	LMMS_STRINGIFY(PLUGIN_NAME),
+	"Oscilloscope",
+	QT_TRANSLATE_NOOP("PluginBrowser", "Oscilloscope plugin to display the incoming audio waveform"),
+	"Keratin <3",
```
Just making totally sure you wanna sign the plugin with this! :smile:

regulus79 replied:
made with love :heart:

bratpeki commented:
> ... which includes a zoomable+scrollable waveform graph...

I just noticed that in the code! That's so cool! So you can pause, zoom with the mouse wheel and move around with the mouse click and drag. Amazing stuff!

sakertooth reviewed (commented):
Quick review. Might take a look a second time but bit preoccupied right now.

bratpeki commented on plugins/Oscilloscope/Oscilloscope.cpp:
```diff
@@ -0,0 +1,88 @@
+	Effect(&oscilloscope_plugin_descriptor, parent, key),
+	m_controls(this)
+{
+	m_ringBuffer = new SampleFrame[BUFFER_SIZE];
```
Is there any benefit to using evident advantages to using `std::array`?

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.h:
```diff
@@ -0,0 +1,92 @@
+	void mouseReleaseEvent(QMouseEvent* me) override;
+	void mouseMoveEvent(QMouseEvent* me) override;
+
+	QColor monoColor() const { return m_monoColor; }
+	void setMonoColor(QColor c) { m_monoColor = c; }
+	QColor leftColor() const { return m_leftColor; }
+	void setLeftColor(QColor c) { m_leftColor = c; }
+	QColor rightColor() const { return m_rightColor; }
+	void setRightColor(QColor c) { m_rightColor = c; }
+	QColor minorLineColor() const { return m_minorLineColor; }
+	void setMinorLineColor(QColor c) { m_minorLineColor = c; }
+	QColor clippingLineColor() const { return m_clippingLineColor; }
+	void setClippingLineColor(QColor c) { m_clippingLineColor = c; }
+	QColor backgroundColor() const { return m_backgroundColor; }
+	void setBackgroundColor(QColor c) { m_backgroundColor = c; }
```
Honestly, getters and setters like this are a waste of space. These should just be public data members, as there is no behavior associated with getting or setting the data.

Assuming we are keeping these getters and setters (if it works, it works, regardless of OOP nonsense), the formatting is fine as-is. From the [coding conventions](https://github.com/LMMS/lmms/wiki/Coding-Conventions):

```cpp
// If the block can fit on one line, it is acceptable to format it as shown below.
// (Note that rules 1 and 2 do not apply, the contents are separated by a space on either side.)
if (m_sample > 0) { --m_sample; }
```

Relevant comment from @tresf in the [Discord](https://discord.com/channels/203559236729438208/784594576223633478/1338606718422356030)

> I think that -- in general -- getters and setters are a pretty big space waster in code, so any convention that makes them easier to scroll past is preferred.  That said, I normally code in Java and over there, getters and setters are so darn bloated in class files that the IDEs just collapse them for you.  I know not everyone uses an IDE and that not all IDEs do this by default, but I wanted to share my experience that the smaller the better. 

Even if you are asking for horizontal alignment, I don't think that's an improvement.

```cpp
	QColor monoColor() const              { return m_monoColor; }
	void   setMonoColor(QColor c)         { m_monoColor = c; }
	QColor leftColor() const              { return m_leftColor; }
	void   setLeftColor(QColor c)         { m_leftColor = c; }
	QColor rightColor() const             { return m_rightColor; }
	void   setRightColor(QColor c)        { m_rightColor = c; }
	QColor minorLineColor() const         { return m_minorLineColor; }
	void   setMinorLineColor(QColor c)    { m_minorLineColor = c; }
	QColor clippingLineColor() const      { return m_clippingLineColor; }
	void   setClippingLineColor(QColor c) { m_clippingLineColor = c; }
	QColor backgroundColor() const        { return m_backgroundColor; }
	void   setBackgroundColor(QColor c)   { m_backgroundColor = c; }
```

Let's just leave these getters and setters alone.

rubiefawn replied:
> If no one is using these getters and setters, then they can be removed. The solution isn't to make them public data members, but to expose only what is necessary. Exposing them as public data members is just as bad as having the getters and setters, maybe even worse, especially if no one is using them.

By "behavior associated with getting or setting the data", I don't mean "unused". I mean that the function returns a naked value with no other computation required; no computation -> no function needed, it's just a value and should be treated like one. I absolutely agree with you that they shouldn't be exposed as `public` (or they should be removed) if they are unused, but that wasn't the point I was originally making.

In fact, I would prefer that the getters and setters be removed entirely where possible for this reason, but advocated for them to be left as-is in order to reduce friction so that this PR doesn't get blocked over what I felt was my own personal preferences.

> I think having consistent formatting and following conventions is the right approach here.

"Consistent formatting" and "following conventions" are two entirely separate things, which both have value, but let's not conflate them. A lot of existing code does not follow our written conventions, and gets updated to follow those conventions as we go. The conventions are the source of truth here, and they are (and continue to be) updated according to developer discussion, not according to how old code is formatted. This is why I presented a citation to the coding conventions in support of my argument.

It follows that if we decide to change the conventions to align with what you are suggesting here, then I will back your requested changes as well. You can see this in my PRs, where the conventions were ambiguous or were modified by vote, and I changed my position to match despite personal preference. Right now, however, there is precedent for compact one-liners, and until that changes, it is what I will advocate for.

> We are generally going to need whitespace anyway if we want documentation in the future, which I am pushing for.

Documentation is good! I am glad you are pushing for its inclusion. The whitespace can be added alongside the documentation when it is added then. The future existence of doc comments is not a compelling reason to add whitespace before adding said documentation.

bratpeki replied:
> Documentation is good! I am glad you are pushing for its inclusion.

Likewise! As someone who's getting into reviewing, comments are a huge time-saver.

regulus79 replied:
Other classes always seem to add these sorts of getter/setter functions for their themeable parameters, so I assumed that they are somehow necessary for Qt to edit the values when the theme is changed. I haven't tested what happens when I remove them.

regulus79 replied:
I tested removing those lines, but it appears that the `Q_PROPERTY` macros above require them to exist. If I also remove the `Q_PROPERTY` lines, those variables seem to be no longer themeable via css.

messmerd replied:
Would this form work?
```cpp
Q_PROPERTY(QColor monoColor MEMBER m_monoColor)
```

See: https://doc.qt.io/qt-6/properties.html

regulus79 commented on plugins/Oscilloscope/Oscilloscope.cpp:
```diff
@@ -0,0 +1,88 @@
+			m_ringBufferIndex = (m_ringBufferIndex + 1) % BUFFER_SIZE;
+		}
+	}
+	return ProcessStatus::ContinueIfNotQuiet;
```
I copied it from Amplifier, so to be honest I'm not exactly sure what the meaning of it is :sweat_smile:

But checking Effect.cpp, it seems it tells the effect to go to sleep if the input sound goes below the gate level? Okay wait maybe that's not what we want... :thinking:

regulus79 committed:
Use less whitespace

Co-authored-by: Fawn <rubiefawn@gmail.com>

regulus79 commented on plugins/Oscilloscope/Oscilloscope.h:
```diff
@@ -0,0 +1,65 @@
+ *
+ */
+
+#ifndef LMMS_TEMPLATE_EFFECT_H
+#define LMMS_TEMPLATE_EFFECT_H
```
Oops, it seems I forgot to have that automatically changed by the template script. I will fix that.

regulus79 committed:
Fix ifndefs

Co-authored-by: Sotonye Atemie <satemiej@gmail.com>

regulus79 committed:
Attempt to replace array with std::array

regulus79 committed:
Always continue, do not sleep when input sound goes below gate level

regulus79 committed:
Attempt to fix build error

regulus79 committed:
Attempt to fix build error again

bratpeki commented:
> Oh no, if you only found out about it by looking at the code, then will most users not realize it exists? /hj

The plugin is immeasurably useful even if you don't know about that, and people who need to pause and look at the waveform would figure it out, I think!

bratpeki commented:
Asking about this line again, LOL! @regulus79, could you talk me through https://github.com/LMMS/lmms/pull/7937/commits/57f07b7f21c9b3cb88ddcf8cad1f1771f27b91d1. Does the removal of the GATE knob done by @messmerd affect that in any way?

regulus79 commented:
> Does the removal of the GATE knob done by @messmerd affect that in any way?

The gate knob should only affect effects which use `ContinueIfNoteQuiet`(I think), so now that I've changed it to `Continue`, it shouldn't depend on the gate knob.

regulus79 committed:
Merge branch 'master' into oscilloscope

regulus79 committed:
Empty initialize array just to be careful

regulus79 committed:
Remove getters/setters and use QProperty MEMBER instead

regulus79 committed:
Merge branch 'master' into oscilloscope

regulus79 committed:
Only put Keratin on copyright line, since saker and bratpeki have expressed that I have changed the file enough that it is appropriate.

sakertooth commented on plugins/Oscilloscope/Oscilloscope.h:
```diff
@@ -0,0 +1,62 @@
+
+	static const int BUFFER_SIZE = 44100*3;
+
+	std::array<SampleFrame, BUFFER_SIZE> buffer() const { return m_ringBuffer; }
```
> The return type could be changed to `std::span<SampleFrame, BUFFER_SIZE>` if a non-owning view is needed.

Is this thread safe? Normally we would need two indices for a SPSC ring buffer, one for the consumer and one for the producer. We can only read multiple elements from the ring buffer safely if we update the indices properly. I don't think simply returning the buffer or a view to the buffer would work here.

regulus79 replied:
Oops, I didn't realize it was copying it by value. I will fix that.

As for thread safety, it's uh.... not :sweat_smile:. There are times were the buffer gets changed by the audio thread before the gui thread gets the updatds frame position, which can cause some minor gui issues (buffers at the very end of the buffer appearing at the start) But it's super minor and only really visible if you are at full zoom-out.

I spent a while trying to think of a way to get it to work correctly, but I'm not sure. Using a mutex would maybe work, but then the audio and gui threads would have to trylock in order not to freeze each other. And then that would mean the buffer might not get updated if the audio thread isn't able to lock the mutex.

Or, maybe it could be done by some kind of double-buffer system where the incoming buffers are saved into one buffer, and every so often those are copied over to the second buffer, which the gui reads from (and only reads from if it's ready). Maybe. I don't know how to do thread safety haha.

Oh wait I just googled SPSC ring buffer. That would be a good idea to have two buffer positions.

Gregaras commented:
[video.webm](https://github.com/user-attachments/assets/4c06d951-807c-40f6-9dbc-d2fc15f16db1)

Maybe the waveform should appear only from the right side? It seems that it doesn't when offset is changed.

bratpeki commented:
> [video.webm](https://github.com/user-attachments/assets/4c06d951-807c-40f6-9dbc-d2fc15f16db1)
> 
> Shouldn't the waveform appear from the right edge of the oscilloscope? (and disappear at the left edge)

In this case, no. You've moved the offset, so the start position is offset. Usually, you won't be touching that knob manually.

regulus79 committed:
Return buffer by reference instead of value

regulus79 committed:
Use std::span

messmerd commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,159 @@
+
+void OscilloscopeGraph::mousePressEvent(QMouseEvent* me)
+{
+	m_mousePressed = true;
```
`m_mousePressed` doesn't appear to be used

messmerd reviewed (commented):
Quick review

regulus79 committed:
Rename ifndef

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

regulus79 committed:
Use auto&& instead of std::function

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

regulus79 committed:
Remove unused m_mousePressed

regulus79 committed:
Update buffer size declaration

regulus79 committed:
Fix style

regulus79 committed:
Fix compile error

regulus79 committed:
Merge branch 'master' into oscilloscope

bratpeki commented:
Apart from a few extra namespaces (I think they're extra, at least, but they might be part of the convention), this looks okay to me. @messmerd can we merge?

tresf commented on plugins/Oscilloscope/Oscilloscope.h:
```diff
@@ -0,0 +1,62 @@
+		return &m_controls;
+	}
+
+	static constexpr int BufferSize = 44100 * 3;
```
Why not address it now?

bratpeki replied:
That and also the fact that the user can just change it and end up clamping the previous number the next time he loads the project. It's gotta be a constant. The question is if this is sufficient.

regulus79 replied:
That is a good point though, half a second isn't a lot. Hmm. I think we picked 3 seconds as a compromise between length and performance. If I remember how the code works correctly, increasing it would only impact performance when the user is zoomed out (I think?). So maybe it would be fine if we bumped it up to 5-10 seconds at 44.1k, which would be like 1.15-2.3 seconds at 192k?

bratpeki replied:
Yeah, 5 sounds good.

rubiefawn replied:
Can't a new buffer with the correct amount of samples for the sample rate be allocated upon plugin creation and when the sample rate changes? Plugin instantiation already involves allocation, and changing sample rate interrupts audio, so allocating a new buffer of the correct size during those events should be fine... right?

regulus79 replied:
The thing is that I'm storing the samples in an `std::array<SampleFrame, BufferSize>`, and in order to define the type, I need to specify the length. Unless we want to use a vector :thinking: I hadn't thought about that actually.

bratpeki replied:
> It should reallocate the buffer to store 5 seconds of audio at that new sample rate.

Yeah, of course, but in case someone had 5 seconds at 192k as the setting, and then switches to 44.1k, we have to clamp the value.

regulus79 replied:
> in case someone had 5 seconds at 192k as the setting, and then switches to 44.1k, we have to clamp the value.

That's also a good point; currently the view size is stored in samples. So if we switched to having a different buffer size depending on the sample rate, maybe it would make sense to store the time instead of the samples.

regulus79 replied:
Should be resolved in the latest commit.

regulus79 committed:
Merge branch 'master' into oscilloscope

regulus79 committed:
Refactor to use LocklessRingBuffer as intermediate between audio and gui

regulus79 committed:
Switch lengthModel to use milliseconds rather than frames

regulus79 commented:
Alright! I finally found some time to return to working on this, and I think I figured out why the `LocklessRingBuffer` was not updating the amount of samples left to read. I had the initialization of the `LocklessRingBufferReader` within the `paintEvent` function, meaning it was being recreated each time, not persisting.

However, once I fixed that and increased the `read_max` amount, it seems to work.

One of the desired benefits of the thread-safe approach was that the jittering of the waveform at high zoom levels would be fixed. I tried testing it for a while to see if I could get it to jitter, and.... well, I think I got it to shift unexpectedly once, and also maybe often imperceptibly jitter in the corner of my eye. I'm not sure if it's been completely fixed, but it's probably better than it was before.

I also reworked the history buffer to be able to be resized depending on the sample rate by having it be an `std::vector<SampleFrame>`. The maximum history length has been increased to five seconds, regardless of sample rate.

The window length model/knob now also works with milliseconds, rather than frames. This required refactoring some things and changing some calculations, so hopefully I didn't break anything with zooming/scrolling etc. The default window length is now 500 ms. Is that too long/short?

NOTE: I have not tested it with different sample rates yet; I do not usually change sample rates ever, so I am unfamiliar with the process in lmms.

In any case, with a bit of testing, I think this PR should be about ready to go!

regulus79 committed:
Attempt to fix waveform barely peeking around left edge when max zoomed out

regulus79 committed:
Increase ring buffer size to fix 192000 Hz sample rate

regulus79 committed:
Read the whole contents of the buffer at once if possible, instead of in quarters (why did SpectrumAnalyzer do it that way?) and reduce buffer size accordingly

regulus79 committed:
Update connect signal/slot syntax

regulus79 committed:
Revert periodicUpdate SIGNAL/SLOT syntax change, due to failing builds

bratpeki commented on plugins/Oscilloscope/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+INCLUDE(BuildPlugin)
+
+BUILD_PLUGIN(oscilloscope Oscilloscope.cpp OscilloscopeControls.cpp OscilloscopeControlDialog.cpp OscilloscopeGraph.cpp MOCFILES OscilloscopeControls.h OscilloscopeControlDialog.h OscilloscopeGraph.h EMBEDDED_RESOURCES logo.svg)
```
@rubiefawn recently made a PR which centralizes the use of `logo.svg` with https://github.com/LMMS/lmms/pull/8124. Could we get that done here, too?

regulus79 committed:
Print warning if ringbuffer is overloaded

regulus79 committed:
Remove logo.svg and use lmms-plugin-logo

regulus79 commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,169 @@
+	int newBufferSize = Engine::audioEngine()->outputSampleRate() * MaxBufferLengthSeconds;
+	m_ringBuffer.resize(newBufferSize);
+	m_ringBufferIndex = m_ringBufferIndex % newBufferSize;
+}
```
Yeah, I'm not sure. I included it because other plugins like LOMM have a `changeSampleRate` slot which updates the sample rate of everything, but I never got it to actually trigger.

regulus79 committed:
Remove extra empty line

regulus79 committed:
Update CMakeLists.txt

bratpeki reviewed (approved):
Approving both in terms of code and testing.

rubiefawn commented on plugins/Oscilloscope/Oscilloscope.h:
```diff
@@ -0,0 +1,63 @@
+	EffectControls* controls() override
+	{
+		return &m_controls;
+	}
```
(Non-blocking) Small getters and setters like this can be [one-liners](https://github.com/LMMS/lmms/wiki/Coding-Conventions#braces):

```suggestion
	EffectControls* controls() override { return &m_controls; }
```

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,168 @@
+	setMinimumSize(400, 200);
+	setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+
+	connect(getGUI()->mainWindow(), SIGNAL(periodicUpdate()), this, SLOT(update()));
```
`SIGNAL()` and `SLOT()` macros should be replaced with the updated syntax used in the line directly after this one.

regulus79 replied:
I tried converting it to the new syntax, but when I do that the builds fail iirc. I think `periodicUpdate` is not visible to plugins or something.

rubiefawn replied:
Strange. I was able to compile this PR with my suggested change—I made sure it worked before suggesting anything specific, otherwise that would have been rather embarrassing for me lol. If you're still getting that compile error, would you mind sharing it?

regulus79 replied:
I was also able to compile it; I just seemed to remember that some of github builds would fail when I did. But it appears like they went fine, so maybe I was misremembering.

regulus79 replied:
Hmm it seems a few of the builds failed.

bratpeki replied:
@rubiefawn's suggestion definitely works!

regulus79 replied:
Are you sure? It seemed like the windows builds for https://github.com/LMMS/lmms/pull/7937/commits/3c29e7cf39c23b0346b204b23762a7e627c5b2d4 failed, complaining about not being able to find the signal or something.

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,168 @@
+	{
+		m_ringBuffer[m_ringBufferIndex] = incomingBuffer[f];
+		m_ringBufferIndex = (m_ringBufferIndex + 1) % m_ringBuffer.size();
+    }
```
A soft tab snuck in!

```suggestion
	for (f_cnt_t f = 0; f < incomingBuffer.size(); ++f)
	{
		m_ringBuffer[m_ringBufferIndex] = incomingBuffer[f];
		m_ringBufferIndex = (m_ringBufferIndex + 1) % m_ringBuffer.size();
	}
```

I was going to suggest replacing this loop with a call to `memcpy()`, but turns out that the ringbuffer implementation does not support anything other than single-member access. Yikes. I suppose that's a problem for later.

rubiefawn commented on plugins/Oscilloscope/Oscilloscope.cpp:
```diff
@@ -0,0 +1,84 @@
+	if (!m_controls.m_pauseModel.value())
+	{
+		// Send the samples from the audio thread over to the gui via a ring buffer; the gui will do all of the processing.
+		f_cnt_t framesWritten = m_inputBuffer.write(buffer, frames);
```
(Non-blocking) `framesWritten` is never mutated and should be declared `const:

```suggestion
		const f_cnt_t framesWritten = m_inputBuffer.write(buffer, frames);
```

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,168 @@
+
+	int windowStartIndex = m_ringBufferIndex + (m_ringBuffer.size() - windowSizeFrames) + (phase) * m_ringBuffer.size();
+	// Quantize start index to prevent flickering (round up to prevent going below the current ring buffer index)
+	int frameOffsetWithinPixel = windowStartIndex % framesPerPixel;
```
(Non-blocking)

- `amp`, `framesPerPixel`, and `frameOffsetWithinPixel` are never mutated and should be declared `const`
- `phase` is used exactly once and can be inlined (it's clear that it's the phase even without this variable binding)
- Some of these lines exceed 120 characters in length, breaking them on operators resolves this

```suggestion
	const float amp = m_controls->m_ampModel.value() * 0.01f;
	int windowSizeFrames = m_ringBuffer.size()
		* m_controls->m_lengthModel.value()
		/ m_controls->m_lengthModel.maxValue();
	const int framesPerPixel = std::max(1, windowSizeFrames / width());

	int windowStartIndex = m_ringBufferIndex
		+ (m_ringBuffer.size() - windowSizeFrames)
		+ m_controls->m_phaseModel.value() * m_ringBuffer.size();
	// Quantize start index to prevent flickering (round up to prevent going below the current ring buffer index)
	const int frameOffsetWithinPixel = windowStartIndex % framesPerPixel;
```

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,168 @@
+	float windowSizeMilliseconds = m_controls->m_lengthModel.value();
+	float phase = m_controls->m_phaseModel.value();
+	float mouseOffset = (1.0f - we->position().x() / width()) * (windowSizeMilliseconds / m_controls->m_lengthModel.maxValue());
+	float zoomAmount = std::clamp(std::exp2(-we->angleDelta().y() / 240.0f), m_controls->m_lengthModel.minValue() / m_controls->m_lengthModel.value(), m_controls->m_lengthModel.maxValue() / m_controls->m_lengthModel.value());
```
(Non-blocking)

- `windowSizeMilliseconds`, `mouseOffset`, and `zoomAmount` are never mutated and should be declared `const`.
- `phase` is used exactly once and can be inlined (it's clear that it's the phase even without this variable binding)
- Some of these lines exceed 120 characters in length, breaking them on operators resolves this

```suggestion
	const float windowSizeMilliseconds = m_controls->m_lengthModel.value();
	const float mouseOffset = (1.0f - we->position().x() / width())
		* (windowSizeMilliseconds / m_controls->m_lengthModel.maxValue());
	const float zoomAmount = std::clamp(
		std::exp2(-we->angleDelta().y() / 240.0f),
		m_controls->m_lengthModel.minValue() / m_controls->m_lengthModel.value(),
		m_controls->m_lengthModel.maxValue() / m_controls->m_lengthModel.value()
	);
```

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,168 @@
+void OscilloscopeGraph::mouseMoveEvent(QMouseEvent* me)
+{
+	float phase = m_controls->m_phaseModel.value();
+	float newPhase = phase + 1.0f * (m_mousePos - me->x()) / width() * (m_controls->m_lengthModel.value() / m_controls->m_lengthModel.maxValue());
```
- What is the purpose of the multiplication by `1.0f` on line 162?
- (Non-blocking) `newPhase` is never mutated and should be declared `const`
- (Non-blocking) `phase` is used exactly once and can be inlined (it's clear that it's the phase even without this variable binding)

```suggestion
	float newPhase = m_controls->m_phaseModel.value() + (m_mousePos - me->x()) / width() * (m_controls->m_lengthModel.value() / m_controls->m_lengthModel.maxValue());
```

regulus79 replied:
I'm guessing the `1.0f` was to prevent integer division from happening when calculating the ratio of the mouse offset pixels to the width of the oscilloscope graph. I will test it locally to see if it works fine without it.

rubiefawn commented on plugins/Oscilloscope/OscilloscopeGraph.h:
```diff
@@ -0,0 +1,90 @@
+	void mousePressEvent(QMouseEvent* me) override;
+	void mouseMoveEvent(QMouseEvent* me) override;
+
+	//! Length of the oscilloscope history in seconds
```
```suggestion
	//! Maximum length of the oscilloscope history in seconds
```

rubiefawn reviewed (commented):
`int` is used in a lot of places to represent a buffer length or index. `int` may be as small as 16 bits and certainly can't represent the maximum possible length of the ring buffer (192000 samples per second * 5 seconds > 65535 samples). I would recommend using at least a `long` instead, which may be no smaller than 32 bits.

regulus79 committed:
Move lmms_plugin_main to first extern C block

Co-authored-by: Fawn <rubiefawn@gmail.com>

regulus79 committed:
Apply suggestions from code review

Co-authored-by: Fawn <rubiefawn@gmail.com>

regulus79 committed:
Cast to float to avoid trunacting to 0 during division

regulus79 committed:
Update colors

Co-authored-by: Fawn <rubiefawn@gmail.com>

regulus79 committed:
Inline stereo variable

Co-authored-by: Fawn <rubiefawn@gmail.com>

regulus79 committed:
Do no save paused state

regulus79 committed:
Remove warnings

regulus79 committed:
Shift the clipping line heights by 1 px to hide them at default zoom/scale

regulus79 commented on plugins/Oscilloscope/Oscilloscope.cpp:
```diff
@@ -0,0 +1,81 @@
+		const f_cnt_t framesWritten = m_inputBuffer.write(buffer, frames);
+		if (framesWritten < frames)
+		{
+			qWarning() << "[Oscilloscope] Frames dropped from audio-gui ringbuffer! Had" << frames << "but only wrote" << framesWritten;
```
Oh wait nevermind, I was mistakenly on 192000 Hz. Should I put it back?

bratpeki commented:
Tested again, looks and feels great!

regulus79 committed:
Merge branch 'master' into oscilloscope

messmerd commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,172 @@
+
+void OscilloscopeGraph::changeSampleRate()
+{
+		const auto newBufferSize = static_cast<std::size_t>(Engine::audioEngine()->outputSampleRate() * MaxBufferLengthSeconds);
```
Looks like this is indented twice

messmerd replied:
```suggestion
	const auto newBufferSize = static_cast<std::size_t>(Engine::audioEngine()->outputSampleRate() * MaxBufferLengthSeconds);
```
removed an extra tab

messmerd commented on plugins/Oscilloscope/OscilloscopeGraph.h:
```diff
@@ -0,0 +1,90 @@
+	LocklessRingBufferReader<SampleFrame> m_inputBufferReader;
+
+	// Buffer which is actually drawn on the screen, not the ring buffer used for communication between threads
+	std::vector<SampleFrame> m_ringBuffer = {};
```
```suggestion
	std::vector<SampleFrame> m_ringBuffer;
```

messmerd commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,172 @@
+{
+		const auto newBufferSize = static_cast<std::size_t>(Engine::audioEngine()->outputSampleRate() * MaxBufferLengthSeconds);
+	m_ringBuffer.resize(newBufferSize);
+	m_ringBufferIndex = m_ringBufferIndex % newBufferSize;
```
If it's possible for this to run concurrently with `paintEvent` (which reads from `m_ringBuffer` and its index), you have a race condition where `paintEvent` tries to read from deallocated memory, causing a crash. The good news is you can use a mutex since this is GUI code.

regulus79 committed:
Move OscilloscopeControls::createView to cpp file to remove include

regulus79 committed:
Remove extra indent

regulus79 committed:
Remove default {} initialization of history buffer

regulus79 committed:
Add std::mutex to prevent buffer from being accessed during sample rate change

regulus79 committed:
Remove SIGNAL/SLOT macro

Co-authored-by: Fawn <rubiefawn@gmail.com>

regulus79 committed:
Revert signal/slot macro removal

bratpeki commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,176 @@
+* License along with this program (see COPYING); if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA.
+*
```
Another extra line. Probably in every source file.

sakertooth commented on plugins/Oscilloscope/OscilloscopeGraph.cpp:
```diff
@@ -0,0 +1,176 @@
+	// Update the ring buffer with any new data from the audio thread
+	auto incomingBuffer = m_inputBufferReader.read_max(effect->inputBuffer().capacity());
+	// Make sure the buffer is not currently being resized by changeSampleRate
+	std::lock_guard<std::mutex> lockGuard(m_bufferMutex);
```
> `// Make sure the buffer is not currently being resized by changeSampleRate`

Why do you need the mutex? AFAICT, `AudioEngine::sampleRateChanged` signal is emitted on the main thread. The audio thread does not emit that signal.

regulus79 replied:
I believe I added it in response to https://github.com/LMMS/lmms/pull/7937#discussion_r2605508890

regulus79 committed:
Merge branch 'master' into oscilloscope

regulus79 committed:
Remove extra lines and update copyright date

regulus79 committed:
Remove isResizable function

regulus79 committed:
Merge branch 'master' into oscilloscope

regulus79 committed:
Remove mutex

Co-authored-by: Dalton Messmer <messmer.dalton@gmail.com>

ewanp2025 commented:
Good work, I understand you have put a lot of time and effort into this. Could I suggest that an X-Y (Lissajous) mode is added to the oscilloscope? This would allow users of Xpressive to create and demonstrate oscilloscope music natively.

bratpeki committed:
Changed fpp to f_cnt_t

bratpeki committed:
Merge branch 'master' into oscilloscope

bratpeki committed:
One fpp_t left unchanged and also removed a whitespace lol

bratpeki commented:
These three commits migrated from the now removed `fpp_t` to `f_cnt_t`.

bratpeki committed:
Moved from Knob::setVolumeKnob to VolumeKnob

bratpeki commented:
This last commit applies a change made in #8253, specifically commit [`1e5ca51bee21...`](https://github.com/LMMS/lmms/pull/8253/changes/1e5ca51bee215099903112652dad33dc611b3daa).
